cirq/sim/clifford/ch_form.py

```diff
@@ -0,0 +1,272 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Union
+import numpy as np
+
+import cirq
+
+
+class CH_Form():
+    r"""
+    A representation of stabilizer states using the CH form,
+
+        $|\psi> = \omega U_C U_H |s>$
+
+    This representation keeps track of overall phase.
+
+    See Bravyi et al, arXiv:1808.00128 for details.
+    """
+
+    def __init__(self, num_qubits, initial_state=0):
+        self.n = num_qubits
+
+        # The state is represented by a set of binary matrices and vectors.
+        # See Section IVa of Bravyi et al
+        self.G = np.eye(self.n, dtype=bool)
+
+        self.F = np.eye(self.n, dtype=bool)
+        self.M = np.zeros((self.n, self.n), dtype=bool)
+        self.gamma = np.zeros(self.n, dtype=int)
+
+        self.v = np.zeros(self.n, dtype=bool)
+        self.s = np.zeros(self.n, dtype=bool)
+
+        self.omega = 1
+
+        def bits(s):
+            while s > 0:
+                yield s & 1
+                s >>= 1
+
+        # Apply X for every non-zero element of initial_state
+        for (i, val) in enumerate(bits(initial_state)):
+            if val:
+                self._X(self.n - i - 1)
+
+    def copy(self):
+        copy = CH_Form(self.n)
+
+        copy.G = self.G.copy()
+        copy.F = self.F.copy()
+        copy.M = self.M.copy()
+        copy.gamma = self.gamma.copy()
+        copy.v = self.v.copy()
+        copy.s = self.s.copy()
+        copy.omega = self.omega
+
+        return copy
+
+    def __repr__(self):
+        """Return the wavefunction representation of the state."""
+        return cirq.dirac_notation(self.to_state_vector())
+
+    def _repr_full(self):
+        """Return the CH form representation of the state. """
+        string = ""
+
+        string += "omega: {:.2f}\n".format(self.omega)
```

Comment:
This is much closer to the `__repr__`. `cirq._compact.proper_repr` can handle numpy arrays.